[popover] Cache auto popover list to avoid unnecessarily iterating the top layer

nt1m opened the pull request:
#### 8141e23838a14f06695a9e7c94c326ca0ceb3c70
<pre>
[popover] Cache auto popover list to avoid unnecessarily iterating the top layer
<a href="https://bugs.webkit.org/show_bug.cgi?id=257427">https://bugs.webkit.org/show_bug.cgi?id=257427</a>
rdar://109932621

Reviewed by Ryosuke Niwa.

hideAllPopoversUntil calls autoPopoverList() and topmostAutoPopover() multiple times, which iterates through the top layer
multiple times. Store another ListHashSet which is a subset of topLayerElements(). Additions/removals are done at the same time
as adding/removing to the top layer, which prevents leaks.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addTopLayerElement):
(WebCore::Document::removeTopLayerElement):
(WebCore::Document::topmostAutoPopover const):
(WebCore::Document::hideAllPopoversUntil):
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Document.h:
(WebCore::Document::autoPopoverList const):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::topmostPopoverAncestor):

Canonical link: <a href="https://commits.webkit.org/264810@main">https://commits.webkit.org/264810@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d3e277c86da135ec5865d2d60f2f03fd0c54ee6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8685 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8975 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10342 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8694 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10963 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8946 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11559 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8831 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9846 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10499 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7146 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7942 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15459 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/8263 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11433 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8572 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7838 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2111 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12050 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8317 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8984,16 +8984,30 @@ void Document::keyframesRuleDidChange(const String& name)
+        if (candidatePopover->hasFullscreenFlag())
+            return;
+#endif
+        if (is<HTMLDialogElement>(candidatePopover) && downcast<HTMLDialogElement>(candidatePopover)->isModal())
```
Should use `*candidatePopover` here.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8984,16 +8984,30 @@ void Document::keyframesRuleDidChange(const String& name)
+    if (auto* candidatePopover = dynamicDowncast<HTMLElement>(element); candidatePopover && candidatePopover->isPopoverShowing() && candidatePopover->popoverState() == PopoverState::Auto) {
+        RELEASE_ASSERT(m_autoPopoverList.contains(*candidatePopover));
+        auto didRemove = m_autoPopoverList.remove(*candidatePopover);
+        RELEASE_ASSERT(didRemove);
```
This `RELEASE_ASSERT` and the one above are doing the same check. We don’t need both.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8984,16 +8984,30 @@ void Document::keyframesRuleDidChange(const String& name)
     auto didRemove = m_topLayerElements.remove(element);
     RELEASE_ASSERT(didRemove);
+    if (auto* candidatePopover = dynamicDowncast<HTMLElement>(element); candidatePopover && candidatePopover->isPopoverShowing() && candidatePopover->popoverState() == PopoverState::Auto) {
+        RELEASE_ASSERT(m_autoPopoverList.contains(*candidatePopover));
```
This release assert would do a secondary lookup. Use find, which returns an iterator, so that the value can be used to remove the item instead doing a secondary hashmap lookup.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8984,16 +8984,30 @@ void Document::keyframesRuleDidChange(const String& name)
+    if (auto* candidatePopover = dynamicDowncast<HTMLElement>(element); candidatePopover && candidatePopover->isPopoverShowing() && candidatePopover->popoverState() == PopoverState::Auto) {
+        RELEASE_ASSERT(m_autoPopoverList.contains(*candidatePopover));
+        auto didRemove = m_autoPopoverList.remove(*candidatePopover);
+        RELEASE_ASSERT(didRemove);
```
This release assert is somewhat redundant. Alternatively, we can just keep this release assert and remove the other one.

nt1m replied:
Removed the other assert.

darinadler reviewed (commented):
Was working on review and had comments. One is a duplicate.

nt1m committed:
[popover] Cache auto popover list to avoid unnecessarily iterating the top layer
https://bugs.webkit.org/show_bug.cgi?id=257427
rdar://109932621

Reviewed by Ryosuke Niwa.

hideAllPopoversUntil calls autoPopoverList() and topmostAutoPopover() multiple times, which iterates through the top layer
multiple times. Store another ListHashSet which is a subset of topLayerElements(). Additions/removals are done at the same time
as adding/removing to the top layer, which prevents leaks.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addTopLayerElement):
(WebCore::Document::removeTopLayerElement):
(WebCore::Document::topmostAutoPopover const):
(WebCore::Document::hideAllPopoversUntil):
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Document.h:
(WebCore::Document::autoPopoverList const):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::topmostPopoverAncestor):

Canonical link: https://commits.webkit.org/264810@main
